src/jdk.security.auth/share/classes/com/sun/security/auth/module/GssLoginModule.java

```diff
@@ -0,0 +1,656 @@
+                System.out.println("\t\t[GssLoginModule] acquired" +
+                    " acceptor credentials");
+        }
+        if (gssName == null && gssICred != null)
+            gssName = gssICred.getName();
+        if (gssName == null && gssACred != null)
+            gssName = gssACred.getName();
+    }
+
+    private void attemptAuthentication(boolean getPasswdFromSharedState)
+        throws LoginException {
+
+        // Get a name, maybe
+        if (name == null) {
+            if (useDefaultCreds) {
+                try {
+                    getcreds();
+                    return;
+                } catch (GSSException e) {
+                    throw new LoginException(e.getMessage());
+                }
+            }
+            if (tryDefaultCreds) {
+                try {
+                    getcreds();
+                    return;
+                } catch (GSSException e) { }
+            }
+
+            promptForName(getPasswdFromSharedState);
+            if (name == null)
+                throw new LoginException ("Unable to determine a GSS name");
+        }
+
+        try {
+            gssName = manager.createName(name, nametypeOid);
+        } catch (GSSException e) {
+            throw new LoginException ("Unable to import GSS name");
+        }
+
+        promptForPass(getPasswdFromSharedState);
+
+        try {
+            getcreds();
+        } catch (GSSException e) {
+            throw new LoginException(e.getMessage());
+        }
+    }
+
+    private void promptForName(boolean getPasswdFromSharedState)
+        throws LoginException {
+        if (getPasswdFromSharedState) {
+            // use the name saved by a module earlier in the stack
+            name = (String)sharedState.get(NAME);
+            if (name == null || name.length() == 0)
+                name = defName;
+            if (debug) {
+                System.out.println("\t\t[GssLoginModule] username from" +
+		    " shared state is " + name);
+            }
+            if (name != null && name.length() > 0)
+                return;
+        }
+
+        if (doNotPrompt)
+            return; // name may be null
+
+        if (callbackHandler == null)
+            throw new LoginException("No CallbackHandler "
+                                     + "available "
+                                     + "to prompt for authentication "
+                                     + "information from the user");
+
+        try {
+            String defUsername = System.getProperty("user.name");
+
+            Callback[] callbacks = new Callback[1];
+            MessageFormat form = new MessageFormat(
+                                   getAuthResourceString(
+                                   "GSS.name.defName."));
+            Object[] source =  {defUsername};
+            callbacks[0] = new NameCallback(form.format(source));
+            callbackHandler.handle(callbacks);
+            name = ((NameCallback)callbacks[0]).getName();
```

Comment:
This feels needlessly roundabout. Just store the callback in an appropriately typed local.